tdm/docs/architecture.md

```diff
@@ -0,0 +1,261 @@
+implementations. Intel TDX Connect or AMD SEV-TIO, for example, rely on device
+manufacturers to support the TDISP, SPDM and PCIe IDE protocols to give
+confidential guests access to trusted MMIO and secure DMA between them and
+devices which trustworthiness is verifiable.
+
+The TEE-IO security model is built on top of this hardware support but also on
+the ability for confidential guests to ultimately decide if they accept or
+reject TEE-IO compliant devices into their TCB. The TDM, as a confidential guest
+component, is the TCB acceptance decision maker and is responsible for notifying
+the underlying Trusted Security Manager (TSM) about its decision. By accepting a
+TDI into the guest TCB, the TDM explicitly requests the TSM to move the device
+into an operational state. The TDM builds the acceptance decision after having
+verified and authenticated the TDI. It must also attest of the assigned device
+trustworthiness, through a local or remote attestation process.
+
+## Scope
+
+## Architectural Overview
+
+High-level view:
+
+```
+
+
+                                   ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+                                   ┃                            Confidential Guest ┃
+                                   ┃           ╔═══════════════════════╗           ┃
+                                   ┃           ║Trusted Device Manager ║           ┃
+                                   ┃           ║                       ║           ┃
+                                   ┃           ║┌─────────────────────┐║           ┃
+                                   ┃           ║│ Attestation Module  │║           ┃
+                                   ┃      ┌───▷║│                     │║           ┃
+                                   ┃      │    ║└─────────────────────┘║           ┃
+                                   ┃      │    ║                       ║           ┃
+                                   ┃      │    ║ ╔════════╗ ╔════════╗ ║           ┃
+                                   ┃      │    ╚═╣  TEE   ╠═╣  TEE   ╠═╝           ┃
+                                   ┃      │      ║ Plugin ║ ║ Plugin ║             ┃
+                                   ┃     udev    ╚════════╝ ╚════════╝             ┃
+                                   ┃ (PCIe device                △                 ┃
+                                   ┃  detection)                 │                 ┃
+                                   ┃      │               Device Evidence          ┃
+                                   ┃      │                Accept/Reject           ┃
+                                   ┃      │                      │                 ┃
+                                   ┃      │                      ▽                 ┃
+                                   ┃      │                  ┌───────┐             ┃
+                                   ┃      │  ┌───────────────┤TEE ABI├────┐        ┃
```

Comment:
I accept that there are cases where userspace needs an override for the evidence acceptance flow especially in these early days where there might not be unification across devices on evidence formats, but I want to advocate for and end goal where the kernel only needs a certificate chain to be able to make acceptance decisions. I am reluctant to sign-off on a design direction that makes the kernel beholden to userspace for device usage as I worry about scenarios like error recovery and power management where userspace upcalls are not feasible.